torch/utils/_pytree.py

```diff
@@ -597,14 +597,32 @@ def tree_structure(tree: PyTree) -> TreeSpec:
     return tree_flatten(tree)[1]
 
 
-def tree_map(func: Callable[..., Any], tree: PyTree) -> PyTree:
-    flat_args, spec = tree_flatten(tree)
-    return tree_unflatten([func(i) for i in flat_args], spec)
+def tree_map(func: Callable[..., Any], tree: PyTree, *rest: PyTree) -> PyTree:
+    """Maps a multi-input function over pytree args to produce a new pytree.
 
+    Args:
+      func: function that takes ``1 + len(rest)`` arguments, to be applied at the
+        corresponding leaves of the pytrees.
+      tree: a pytree to be mapped over, with each leaf providing the first
+        positional argument to ``func``.
+      rest: a tuple of pytrees, each of which has the same structure as ``tree``
+        or has ``tree`` as a prefix.
+
+    Returns:
+      A new pytree with the same structure as ``tree`` but with the value at each
+      leaf given by ``f(x, *xs)`` where ``x`` is the value at the corresponding
+      leaf in ``tree`` and ``xs`` is the tuple of values at corresponding nodes in
+      ``rest``.
 
-def tree_map_(func: Callable[..., Any], tree: PyTree) -> PyTree:
-    flat_args = tree_leaves(tree)
-    deque(map(func, flat_args), maxlen=0)  # consume and exhaust the iterable
+    """
+    flat_args, specs = zip(*(tree_flatten(tree) for tree in (tree, *rest)))
+    spec = specs[0]
```

Comment:
Should we check all the `spec`s are the same?

Reply:
Not sure, Jax's contract is 
> rest (Any) – a tuple of pytrees, each of which has the same structure as tree or has tree as a prefix. *
They don't specify that the specs must match.

I can imagine a case where the structure is the same, but the spec is different (for some custom PyTree that carries some metadata). 

This actually makes me realize that Jax's example won't work here:

```python
jax.tree_util.tree_map(lambda x, y: [x] + y, [5, 6], [[7, 9], [1, 2]])
[[5, 7, 9], [6, 1, 2]]
```

Here we fully flatten all trees, which isn't compatible with this example :/

Reply:
I opened a PR to support this.

- #115642